# Language Modeling using Ngram

In this Exercise, you are going to use NLTK which is a natural language processing library for python to create a bigram language model and its variation. You will build one model for each of the following type and calculate their perplexity:
- Unigram Model
- Bigram Model
- Bigram Model with Laplace smoothing
- Bigram Model with Interpolation

As a reminder,
### Don't forget to shut down your instance on Gcloud when you are not using it ###


In [1]:
#First we import necessary library such as math, nltk, bigram, and collections.
import math
import nltk
import io
import random
from random import shuffle
from nltk import bigrams, trigrams
from collections import Counter, defaultdict
random.seed(999)

BEST2010 is a free Thai NLP dataset by NECTEC usually use as a standard benchmark for various NLP tasks includeing language modeling. BEST2010 is separated into 4 domain article, encyclopedia, news and novel. The data is already  tokenized using '|' as a separator.

For example,

ตาม|ที่|นางประนอม ทองจันทร์| |กับ| |ด.ช.กิตติพงษ์ แหลมผักแว่น| |และ| |ด.ญ.กาญจนา กรองแก้ว| |ป่วย|สงสัย|ติด|เชื้อ|ไข้|ขณะ|นี้|ยัง|ไม่|ดี|ขึ้น|

In [2]:
# We choose news domain as our dataset
best2010=[]
fp= io.open('BEST2010/news.txt','r',encoding='utf-8')
for i,line in enumerate(fp):
    best2010.append(line.strip()[:-2])
fp.close()
vocabuary =set()
total_word_count =0
for line in best2010:
    for word in line.split('|'):        
        vocabuary.add(word)
        total_word_count+=1

In [3]:
#For simplicity, we assumes that each line is a sentence.
print ('Total sentences in BEST2010 news dataset :\t'+ str(len(best2010)))
print ('Total word counts in BEST2010 news dataset :\t'+ str(total_word_count))
print ('Total vocabuary in BEST2010 news dataset :\t'+ str(len(vocabuary)))

Total sentences in BEST2010 news dataset :	30969
Total word counts in BEST2010 news dataset :	1660188
Total vocabuary in BEST2010 news dataset :	38605


We separate out input into 2 sets, train and test data with 70:30 ratio

In [4]:
sentences = best2010
# The data is separated to train and test set with 70:30 ratio.
train = sentences[:int(len(sentences)*0.7)]
test = sentences[int(len(sentences)*0.7):]

# We will use 1/total word count as a default value for unknown word
unk_value = math.pow(len(vocabuary),-1)

# Unigram

In this section, we will demonstrate how to build a unigram language model

In [5]:
def getUnigramModel(data):
    model = defaultdict(lambda: 0)
    total_word_count =0
    for sentence in data:
        sentence +=  u'|</s>'
        for w1 in sentence.split('|'):
            model[w1] +=1.0
            total_word_count+=1
    for w1 in model:
        model[w1] = model[w1]/(total_word_count)
    return model

In [6]:
model = getUnigramModel(train)

In [7]:
def getLnValue(x):
    if x >0.0:
        return math.log(x)
    else:
        return math.log(unk_value)

In [8]:
#problability of 'นายก'
print(getLnValue(model[u'นายก']))
#for example, problability of 'นายกรัฐมนตรี' which is an unknown word is equal to
print(getLnValue(model[u'นายกรัฐมนตรี']))
#problability of 'นายก' 'ได้' 'ให้' 'สัมภาษณ์' 'กับ' 'สื่อ'
prob = getLnValue(model[u'นายก'])+getLnValue(model[u'ได้'])+ getLnValue(model[u'ให้'])+getLnValue(model[u'สัมภาษณ์'])+getLnValue(model[u'กับ'])+getLnValue(model[u'สื่อ'])
print ('Problability of a sentence', math.exp(prob))

-6.55547996498
-10.5611370807
('Problability of a sentence', 2.601573351867736e-16)


# Perplexity

In order to compare language model we need to calculate perplexity. In this task you should write a perplexity calculation code for the unigram model. The result perplexity should be around 553.08 and
480.49 on train and test data.

## TODO #1 Calculate perplexity

In [9]:
def calculate_sentence_ln_prob(sentence, model):
    word = sentence.split('|')
    ln_prob = .0
    for i in range(0,len(word)-1):
        ln_prob +=(getLnValue(model[word[i]]))
    return ln_prob

def perplexity(test,model):
    ln_prob  = .0
    word_count = 0
    for sentence in test:
        sentence += u'|</s>'
        word_count += len(sentence.split('|'))
        ln_prob    += calculate_sentence_ln_prob(sentence,model)
    return math.exp(-ln_prob/word_count)

In [10]:
print(perplexity(train,model))
print(perplexity(test,model))

553.082183054
480.492770536


## TODO #2 Please explain why this model give us such a high perplexity.

**Your answer**:  Write Your Answer here

# Bigram

Next, you will create a better language model than a unigram (which is not much to compare with). But first, it is very tedious to count every pair of words that occur in our corpus by ourselves. In this case, nltk provide us a simple library which will do it for us.

In [11]:
#example of nltk usage for bigram
sentence = 'I always search google for an answer .'

print('This is how nltk generate bigram.')
for w1,w2 in bigrams(sentence.split(), pad_right=True, pad_left=True):
    print (w1,w2)
print('None is used as a start and end of sentence symbol.')

This is how nltk generate bigram.
(None, 'I')
('I', 'always')
('always', 'search')
('search', 'google')
('google', 'for')
('for', 'an')
('an', 'answer')
('answer', '.')
('.', None)
None is used as a start and end of sentence symbol.


Now, you should be able to implement a bigram model by yourself. Also, you must create a new perplexity calculation for bigram. The result perplexity should be around 57.07 and 155.98 on train and test data.

## TODO #3 Write Bigram Model

In [12]:
def getBigramModel(data):
    model = defaultdict(lambda: defaultdict(lambda: 0.0))
    for sentence in data:
        for w1, w2 in bigrams(sentence.split('|'), pad_right=True, pad_left=True): #None I go to school . None
            model[w1][w2] +=1.0
    for w1 in model:
        unigram = float(sum(model[w1].values()))
        for w2 in model[w1]:
            model[w1][w2] = model[w1][w2]/unigram
    return model
model = getBigramModel(train)

## TODO #4 Write Perplexity for Bigram Model

In [13]:
def perplexity(data,model):
    # Write Your Code here
    pass

In [14]:
print (perplexity(train,model) )
print (perplexity(test, model))

57.0715274828
155.975300081


# Smoothing

Usually any ngram models have a sparsity problem, which means it does not have every possible ngram of words in the dataset. Smoothing techniques can alleviate this problem. In this section, you will implement two basic smoothing methods laplace smoothing and interpolation for bigram.

## TODO #5 write Bigram with Laplace smoothing (Add-One Smoothing)

In [ ]:
#Laplace Smoothing
def getBigramWithLaplaceSmoothing(data):
    pass

In [28]:
model =getBigramWithLaplaceSmoothing(train)
print (perplexity(train,model) )
print (perplexity(test, model))

1040.81596701
206.733479856


## TODO #6 Write Bigram with Interpolation
lambda value is 0.7 for bigram, 0.25 for unigram, and 0.05 for unknown word

In [ ]:
#interpolation
def getBigramWithInterpolation(data):
    pass

In [16]:
model = getBigramWithInterpolation(train)
print (perplexity(train,model))        
print (perplexity(test,model))

38.809248557
110.076583883


the result perplexity on training and testing should be 

    1040.82 206.73 for Laplace smoothing
    38.81, 110.08 for Interpolation

# Language modeling on multiple domains

Sometimes, we do not have enough data to create a language model for a new domain. In that case, we can improvised by combining several models to improve result on the new domain.

In this exercise you will try to merge two language model from news and article domains to create a language model for the encyclopedia domain.

In [17]:
# create article data
encyclo_data=[]
fp= io.open('BEST2010/encyclopedia.txt','r',encoding='utf-8')
for i,line in enumerate(fp):
    encyclo_data.append(line.strip()[:-2])
fp.close()

First, you should try to calculate perplexity of your bigram with interpolation on article data. The result perplexity should be around 462.68. 

For your information, a bigram model with interpolation using ariticle data to test on encyclopedia data has a perplexity of 278.20

In [18]:
#print perplexity of bigram with interpolation on article data        
print (perplexity(encyclo_data,model))

462.676320738


## TODO #7 
Write a model that produce 250.0 or less perplexity on encyclopedia data without using data from the encyclopedia as training data. (Hint : Try to combine a model with news data and a model with article data together.)

In [ ]:
#Write your answer here